Emoji reactions support, take 4

TheEssem opened the pull request:
Replaces #1980, #2127, and #2221.

All previous commits have been merged and rebased. I've also fixed a couple of formatting issues and inconsistencies.

Let me know if there's anything missing. I hope to have this PR serve as a base for continued development of this patch.

TheEssem committed:
Add support for emoji reactions
Squashed, modified, and rebased from glitch-soc/mastodon#2221.

Co-authored-by: fef <owo@fef.moe>
Co-authored-by: Jeremy Kescher <jeremy@kescher.at>
Co-authored-by: neatchee <neatchee@gmail.com>
Co-authored-by: Ivan Rodriguez <104603218+IRod22@users.noreply.github.com>
Co-authored-by: Plastikmensch <plastikmensch@users.noreply.github.com>

TheEssem committed:
Add support for emoji reactions
Squashed, modified, and rebased from glitch-soc/mastodon#2221.

Co-authored-by: fef <owo@fef.moe>
Co-authored-by: Jeremy Kescher <jeremy@kescher.at>
Co-authored-by: neatchee <neatchee@gmail.com>
Co-authored-by: Ivan Rodriguez <104603218+IRod22@users.noreply.github.com>
Co-authored-by: Plastikmensch <plastikmensch@users.noreply.github.com>

TheEssem commented:
I've been thinking about how showing who reacted to a post could work on the web UI; would it be better to show users like in the favorites/boosts lists in chronological order but with the emoji next to them somewhere (maybe next to the follow button?), or would it be better to have a list with separate tabs for different emojis?

Here are a couple of mockups for a potential button to view them; one option I thought of was to put a clickable smiley icon next to the boost/favorite counts (taken with upstream's Material Icons set, but probably doable with FA4):
![image](https://github.com/glitch-soc/mastodon/assets/21112968/d8ba1d6b-05d8-4307-bc26-daa6abd83921)
Maybe an "info" button can go at the end of the reactions instead (though likely slightly larger than depicted here:
![image](https://github.com/glitch-soc/mastodon/assets/21112968/7a68af5c-91fd-4bf4-9983-c29af9e6759d)

Plastikmensch commented:
> I've been thinking about how showing who reacted to a post could work on the web UI; would it be better to show users like in the favorites/boosts lists in chronological order but with the emoji next to them somewhere (maybe next to the follow button?), or would it be better to have a list with separate tabs for different emojis?
> 
> Here are a couple of mockups for a potential button to view them; one option I thought of was to put a clickable smiley icon next to the boost/favorite counts (taken with upstream's Material Icons set, but probably doable with FA4): ![image](https://user-images.githubusercontent.com/21112968/281968435-d8ba1d6b-05d8-4307-bc26-daa6abd83921.png) Maybe an "info" button can go at the end of the reactions instead (though likely slightly larger than depicted here: ![image](https://user-images.githubusercontent.com/21112968/281969255-7a68af5c-91fd-4bf4-9983-c29af9e6759d.png)

See: https://github.com/glitch-soc/mastodon/pull/2221#issuecomment-1586109183

Also Font Awesome will likely be removed.

MagicLike commented:
> > I've been thinking about how showing who reacted to a post could work on the web UI; would it be better to show users like in the favorites/boosts lists in chronological order but with the emoji next to them somewhere (maybe next to the follow button?), or would it be better to have a list with separate tabs for different emojis?
> > 
> > Here are a couple of mockups for a potential button to view them; one option I thought of was to put a clickable smiley icon next to the boost/favorite counts (taken with upstream's Material Icons set, but probably doable with FA4): ![image](https://user-images.githubusercontent.com/21112968/281968435-d8ba1d6b-05d8-4307-bc26-daa6abd83921.png) Maybe an "info" button can go at the end of the reactions instead (though likely slightly larger than depicted here: ![image](https://user-images.githubusercontent.com/21112968/281969255-7a68af5c-91fd-4bf4-9983-c29af9e6759d.png)
> 
> See: https://github.com/glitch-soc/mastodon/pull/2221#issuecomment-1586109183
> 
> Also Font Awesome will likely be removed.

Wouldn't it be better for UX and general usability to do it "Misskey-Style" and show the usernames on hover? The list could be added anyway for a better mass-overview.

imdatsolak commented:
If you do it MissKey-style (hover on mouse-over), how would you do that on mobile devices, where there is no mouse?

mkody commented:
Misskey and derivatives use both a mouse-over for a quick peek (shows only 11 people and then a "+[n]") and have either a link in the post's menu or a tab to list again all reactions and who used it, similarly to how Discord does it.

<details>
<summary>Screenshots</summary>

Mouse-over:
![Mouse-over a reaction, showing the popover with 11 accounts who reacted with the first emote](https://github.com/glitch-soc/mastodon/assets/843787/d7c1c627-9b8e-409e-bcc1-0ce2d217b6a1)

Misskey's variant:
![Misskey's "Reactions" element in the post's menu](https://github.com/glitch-soc/mastodon/assets/843787/cf83eb43-6cc4-4d24-9295-3d9d488c1e3a) &nbsp; ![Misskey's reactions popup](https://github.com/glitch-soc/mastodon/assets/843787/c6c8cd35-e3a3-4687-9615-753cf443b559)

And on Firefish:
![Firefish's reactions tab](https://github.com/glitch-soc/mastodon/assets/843787/bb00e834-6eeb-42c5-bf8c-4845257978df)

</details>

Plastikmensch commented:
I'm still for removing modifications of the vanilla flavour.
It was frankly a mistake to suggest and doubles the workload of maintaining this beast.

TheEssem committed:
Add notification emails for reactions

TheEssem committed:
Add notification emails for reactions

TheEssem commented:
> Also this modifies locales managed by Crowdin again

Ah, apologies; I didn't realize that's why they were removed. Fixed.

TheEssem committed:
Fix reblog reactions

TheEssem committed:
Fix reblog reactions

jippi commented on app/mailers/notification_mailer.rb:
```diff
@@ -6,8 +6,8 @@ class NotificationMailer < ApplicationMailer
   before_action :process_params
-  before_action :set_status, only: [:mention, :favourite, :reblog]
-  before_action :set_account, only: [:follow, :favourite, :reblog, :follow_request]
+  before_action :set_status, only: [:mention, :favourite, :reaction, :reblog]
```
From a server owner cost management and email reputation perspective, sending an email per reaction could get incredibly expensive really fast.

It would be great to have an option for

1. roll reactions these up into 30-60min aggregate mails
2. a server setting to outright disable email notifications for reactions

Personally, I would use (2) on my server, simply because the cost aspect of potentially sending 100s of mails to a user is not something I'm interested in bankrolling on behalf of my users, and it can have significant email deliverability impact as well if users get spammed and never opens/interacts with them.

TheEssem replied:
Reaction email notifications are disabled by default in user settings and notification emails of any type are only sent to inactive users anyways (unless a specific setting is manually enabled).

jippi replied:
It's great its disabled by default, but I do think it would be great to have a server setting to turn it off, with no way to turn it on - I really don't want to see someone (myself included) being burned by potentially expensive transactional mails due to malice of some trolls, or some random toot going viral and burning through mail credits.

It can be quite expensive for on-demand transactional emails, and can prohibit verification mails from going out if the server admin is using a pre-paid delivery plan with X mails per month.

Plastikmensch replied:
While the concern is understandable, this is not something this PR can or even should address.
As already stated, Mastodon rarely even sends e-mail notifications by default and most people don't enable those.
This just adds the option to receive e-mails for reactions, in line with other notifications.
Concerns about potentially burning e-mail credit should be brought up upstream.

TheEssem committed:
Refactor react services

TheEssem committed:
Refactor react services

TheEssem commented:
Once the Material Symbols set is ported, I'm thinking that the current "plus" icon could be replaced with the set's own "Add Reaction" icon to better indicate what it's for and bring it more in line with other UI implementations.
![image](https://github.com/glitch-soc/mastodon/assets/21112968/012b2649-4dfb-4422-8222-76398f4c75c1)

ShadowJonathan commented:
@TheEssem what're the blockers for un-drafting this, atm, btw?

TheEssem commented:
> @TheEssem what're the blockers for un-drafting this, atm, btw?

I'm mainly just waiting for further review.

ShadowJonathan commented:
If you mean maintainer review then I'd suggest making it non-draft, since else (ime) maintainers won't look at it

ClearlyClaire commented:
I'm sorry I really don't have much mental bandwidth right now, and my priority is bringing glitch-soc to a more maintainable state. There's a lot I've postponed, and it's becoming a lot.

lunaisnotaboy commented:
Is this suitable for being merged into a fork?

lunaisnotaboy reviewed (approved):
LGTM

lunaisnotaboy commented:
Running in production at [`michiru.me`](https://michiru.me), and it works great!

jordemort commented:
Hm, I applied this to my fork but it doesn't seem to have taken. I've got MAX_REACTIONS set to 32 in .env.production but I don't see the `+` button to add them to anything.

gBasil commented:
> Hm, I applied this to my fork but it doesn't seem to have taken. I've got MAX_REACTIONS set to 32 in .env.production but I don't see the `+` button to add them to anything.

Shot in the dark, are you using the Glitch frontend flavor or the Vanilla Mastodon one?

jordemort commented:
> Shot in the dark, are you using the Glitch frontend flavor or the Vanilla Mastodon one?

I had that thought as well, but it's definitely the Glitch frontend

lunaisnotaboy commented:
> > Shot in the dark, are you using the Glitch frontend flavor or the Vanilla Mastodon one?
> 
> I had that thought as well, but it's definitely the Glitch frontend

Did you rebuild the assets?

jordemort commented:
> Did you rebuild the assets?

Yes, that was it; I'm running in Docker and doing a little dance before container startup to move assets over to a volume so nginx can serve them directly and I had `public/assets` but not `public/packs` updated; my problem was self-inflicted and it's all sorted now.

emilweth commented:
Is there any way to get reaction details the same way you can have the list of people who boosted/liked a post?

shleeable commented:
Can this be moved to mainline? or does it depend on a glitch functionality?

ShadowJonathan commented:
Moving this PR to mainline would take substantial more time and review, it's easier to let it land and mature on glitch-soc, and then to port back to main after a while, I remember a bunch of other features get the same treatment

TheEssem committed:
Linting fixes

TheEssem committed:
Linting fixes

ShadowJonathan commented on app/serializers/rest/reaction_serializer.rb:
```diff
@@ -21,11 +21,25 @@ def custom_emoji?
     object.custom_emoji.present?
   end
 
+  def name
```
note: since this is editing the reaction serializer (that is also used for announcements), could you verify that it doesnt break or bug the announcement reaction list?

TheEssem replied:
They seem good here; this announcement was made with the patch applied.
![image](https://github.com/glitch-soc/mastodon/assets/21112968/8dfc7ccc-1943-4428-af6d-4b5bbefafaf6)

TheEssem committed:
Add reaction notification column settings
This was in a previous PR. Not quite sure how it didn't carry over.

TheEssem committed:
Add reaction notification column settings
This was in a previous PR. Not quite sure how it didn't carry over.

TheEssem committed:
Fix rubocop complaint

TheEssem committed:
Fix rubocop complaint

Johann150 commented on app/lib/activitypub/activity/like.rb:
```diff
@@ -3,12 +3,38 @@
   end
+
+  # Misskey delivers reactions as likes with the emoji in _misskey_reaction
+  # see https://misskey-hub.net/ns.html#misskey-reaction for details
```
This only checks the `_misskey_reaction` attribute (defined by Misskey), even though the `content` attribute (defined by ActivityPub) has been provided by Misskey [for a long time (since 2020-02-06)](https://github.com/misskey-dev/misskey/commit/988ac80087a116173ab36043a7a9dd2228d2986c#diff-12e8576204846b63bad4472ff3737065ab6985294a984a51bb320628161c2cd2) and [is still provided](https://github.com/misskey-dev/misskey/blob/75034d9240c069baff5a24409ea172374261ea3b/packages/backend/src/core/activitypub/ApRendererService.ts#L274). Not checking the `content` field of a `Like` activity will make reactions from Foundkey not show up correctly.

Since the `content` attribute is defined by ActivityPub it should IMHO be preferred.

( originally posted in https://akkoma.dev/AkkomaGang/akkoma/issues/499 see also https://akkoma.dev/FoundKeyGang/FoundKey/issues/354 )

TheEssem replied:
Should be fixed with 265eeed.

TheEssem committed:
Check for content attribute in Misskey likes

TheEssem committed:
Check for content attribute in Misskey likes

ShadowJonathan commented on app/lib/activitypub/activity/like.rb:
```diff
@@ -3,12 +3,39 @@
+    return if original_status.nil? || !original_status.account.local? || delete_arrived_first?(@json['id'])
 
-    return if original_status.nil? || !original_status.account.local? || delete_arrived_first?(@json['id']) || @account.favourited?(original_status)
+    return if maybe_process_misskey_reaction
```
nit: with the addition of non-misskey react content (generalised in spec), maybe this one could be `maybe_process_embedded_reaction`?

TheEssem replied:
4f15150fa1c71e0db9329c53ed069938852317aa does this.

kescherCode commented:
There may still be a bug present here that could allow a malicious or malfunctioning remote instance to overwrite local or remote emojis, or creates ones that previously did not exist. To reproduce, have three instances federate with each other, each with one post, and each leaving a emoji reaction on every test post (including their own).

Then, each instance reacts with emojis that others reacted with. This results in foreign emojis being created, that don't actually belong to the instance that's in the domain sent to Mastodon, and I have witnessed a local custom emoji being overwritten as well.

My test setup was with the previous iteration of this PR, the latest release of Akkoma, and a fork of Misskey (like Iceshrimp, Sharkey).

kescherCode commented:
Also, reacting locally with remote reactions creates false reactions on remote instances. This only happens with this PR, not other custom emoji reaction implementations.

<details><summary>Screenshots of two emojis that do not belong to the test instance, being listed <strong>under the CDN domain</strong>, not even the instance domain. This may be related to how they are federated out.</summary>

![image](https://github.com/glitch-soc/mastodon/assets/23276523/1a57aec9-4bbb-4049-8618-455410e665fd)
![image](https://github.com/glitch-soc/mastodon/assets/23276523/cffc6ff3-9eb5-4d0e-9335-48a6987fef9f)

</details>

Keep in mind that I have yet to retest this with this iteration of the PR, but I did not want this behaviour to go unnoticed.

kescherCode commented:
Okay, I've tested the issue with this iteration (after finally merging it) and can now safely say: The bug is gone from this one.

TheEssem commented:
Came across a bug, it looks like emojis with presentation sequences (U+FE0F) are considered separate to ones without, which causes the appearance of two different identical reactions:
![image](https://github.com/glitch-soc/mastodon/assets/21112968/0ba621a0-4b86-48b6-8aa1-6f3a993976a8)
Emojis that should have presentation sequences might need to be normalized.

TheEssem committed:
Normalize emojis with variant selectors

TheEssem committed:
Normalize emojis with variant selectors

TheEssem committed:
Make name of like content parser function more general

TheEssem committed:
Make name of like content parser function more general

ShadowJonathan commented on app/services/react_service.rb:
```diff
@@ -0,0 +1,49 @@
+    return unless domain.nil? || status.local?
+
+    normalized = "#{name}\uFE0F"
+    name = normalized if StatusReactionValidator::SUPPORTED_EMOJIS.include?(normalized)
```
The same needs to be applied to the unreact service, and it might break if a server were to send a normalised and unnormalised emoji.

Maybe it's best to move this transform to the API interface?

TheEssem replied:
Thanks for the catch, moved it.

ShadowJonathan replied:
Not quite yet; the fix ([`a82d4e5` (#2462)](https://github.com/glitch-soc/mastodon/pull/2462/commits/a82d4e58eb931159eb3c17573b63a2c8b41f3871)) moves it to catch the `#create` and `#destroy` urls, but the problem was in the display.

I meant to also find a way to merge non-normalised and normalised reacts when requesting reacts per status, and/or also per the notification service, if that's not a huge deal.

The goal is also to prevent duplicate reacts (that a non-normalised and a normalised emoji become 2 of normalised) and etc, but i dont know how much of an effort that is.

ifacodes replied:
just spitballing because i've taken like the most cursory glance at this particular issue and how it's implemented, but can we handle / store emoji objects as potentially containing both a normalized and a non-normalized emoji? and then choose one to prefer to display? I think that would solve the issue how merging theme and preventing duplicate reacts whilst also not messing up create / destroy actions

kescherCode commented:
I would like to suggest the usage of the `heart_plus.svg` icon as the new icon required for this rebase.

kescherCode commented:
(Though we could also use add.svg, as it is used elsewhere already. Still!)

kescherCode commented:
![image](https://github.com/glitch-soc/mastodon/assets/23276523/b76bf636-7426-4738-ab83-b1084e75e719)

In case you want to blindly apply this @TheEssem, simply do a naive rebase without setting a custom emoji and then apply this commit: [Commit view](https://git.kescher.at/CatCatNya/catstodon/commit/f833e4d66d21905204d9799b7a9b75851f5c6639), [Patch file](https://git.kescher.at/CatCatNya/catstodon/commit/f833e4d66d21905204d9799b7a9b75851f5c6639.patch)

kescherCode commented:
![image](https://github.com/glitch-soc/mastodon/assets/23276523/81c155c1-bfd1-4181-a705-58d927730025)

Yes, icons are misaligned, but it seems that's an upstream issue...

Patch: https://git.kescher.at/CatCatNya/catstodon/commit/6ec2c44e5e623fb1c30e982998fd4cb377018a76.patch

kescherCode commented:
Ah, looking at it again, yes, the misaligned *icons* are upstream (reply is definitely not aligned with boost and favourite), but there is an extra div for the emoji reaction button... Updating my "patch for post-rebase" to fix this

kescherCode commented:
Nevermind. Not doing this.

TheEssem commented:
> I would like to suggest the usage of the `heart_plus.svg` icon as the new icon required for this rebase.

I feel like this would be confusing since upstream appears to be changing the favorite/like icon to a heart as well ([as the comments of this seem to indicate](https://github.com/mastodon/mastodon/pull/27385)). Since the new set has a dedicated "Add Reaction" icon that better indicates that it deals with emojis, I think it would be better to use that.

kescherCode commented:
Oh, true. Yeah, go ahead then :D

TheEssem committed:
Quick fixes

TheEssem committed:
Quick fixes

TheEssem committed:
Move reaction normalization to API controller

TheEssem committed:
Move reaction normalization to API controller

TheEssem committed:
Revert variant selector normalization

Probably worth tackling later, but for now it's not worth worrying about; some other implementations (e.g. Misskey's) look to have the same behavior anyways.

TheEssem committed:
Revert variant selector normalization

Probably worth tackling later, but for now it's not worth worrying about; some other implementations (e.g. Misskey's) look to have the same behavior anyways.

TheEssem committed:
Update reaction emails

Reaction icon made by t3rminus@calamity.world

TheEssem committed:
Update reaction emails

Reaction icon made by t3rminus@calamity.world

TheEssem committed:
Simplify reactions API controller

TheEssem committed:
Simplify reactions API controller

TheEssem committed:
Refactor status reactions query

This was done to announcement reactions in 1b0cb3b54d1a1b08922527aefc8135d56d3a1a8d. Might as well do it here too.

TheEssem committed:
Refactor status reactions query

This was done to announcement reactions in 1b0cb3b54d1a1b08922527aefc8135d56d3a1a8d. Might as well do it here too.

TheEssem committed:
Fix rubocop lint issue

TheEssem committed:
Fix rubocop lint issue

Ember-ruby commented:
Reacts from limited instances are shown, like other interactions from limited instances these should be hidden

TheEssem committed:
Purge status reactions on account delete

TheEssem committed:
Purge status reactions on account delete

TheEssem committed:
Hydrate reactions on streaming API

TheEssem committed:
Hydrate reactions on streaming API

puckipedia commented:
Would it be possible to add `"EmojiReact": "litepub:EmojiReact"` to the context extensions, and add it to the context when serializing the emoji reaction? Right now, my server will not reliably accept emoji reactions coming from this PR, as it's not in the context.

kescherCode commented:
![image](https://github.com/glitch-soc/mastodon/assets/23276523/12808bb2-d4d8-4d2e-b79f-3f5f1e3a9347)

Recent merge has misaligned the buttons on origin posts when viewing a reply (in deck and single-column views)

TheEssem commented:
> ![image](https://private-user-images.githubusercontent.com/23276523/304844778-12808bb2-d4d8-4d2e-b79f-3f5f1e3a9347.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwMjc3NzEsIm5iZiI6MTcwODAyNzQ3MSwicGF0aCI6Ii8yMzI3NjUyMy8zMDQ4NDQ3NzgtMTI4MDhiYjItZDRkOC00ZDJlLWI3OWYtM2Y1ZjFlM2E5MzQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE1VDIwMDQzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlmMTE2MWU2YzEzOTMxNWU3YzBkZTVhZDQxNWJkNjJhMThhNGExMzcwY2UwYjlhMjljM2M1YzljNTNjOGEzYjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.30r03dpsp9IPiPDDi2yPobzbpqJr3Z1vwT2CTmWbn7M)
> 
> Recent merge has misaligned the buttons on origin posts when viewing a reply (in deck and single-column views)

I can't seem to reproduce this in deck or single-column, unfortunately...

kescherCode commented:
@TheEssem yeah, sorry about that. It seems I incorrectly applied the changes from the last rebase

TheEssem committed:
Merge fixes

TheEssem committed:
Merge fixes

TheEssem committed:
Fix reaction picker dropdown appearance

TheEssem committed:
Fix reaction picker dropdown appearance

zolexic commented:
i pulled this onto jorts.horse for experimentation (and also just for fun), and can report at least one minor UI headache caused by this feature -- because glitch moves the time/datestamp of an unexpanded post from the top right to the bottom right (instead utilizing the top right for context/meta icons), the addition of a new button along the bottom of a post causes the date/timestamp to be cut off entirely in some contexts. on mobile, for example, the stamp is completely gone for most posts, and on desktop, if timestamp is something involving longer characters ("57m" vs "11m", for example), you'll get "5..." as the stamp. For datestamps, it's even worse -- any posts older than a few days will generally get the first few letters of the month it was posted, and nothing else

i don't see this as a particularly _important_ problem, mind you -- it's a tolerable trade-off for the fun of emoji reacts. but it should be noted as something that probably needs to be tweaked before this could/should be officially merged with main

gmemstr commented:
Might need to be rebased onto master, the `.patch` no longer cleanly applies.

kescherCode committed:
[Glitch+Emoji reactions] Use modern React context for for identity for emoji reaction code

kescherCode committed:
[Glitch+Emoji reactions] Use modern React context for for identity for emoji reaction code

kescherCode commented:
Needs patch similar to https://github.com/CatCatNya/catstodon/commit/47884616f7fd3f7c7cfa95885fb9132e689da354. This is because identity and some other context stuff is now passed via props instead of the old context.

kescherCode commented:
Perhaps the commit that adds the identity consumer should be removed from the history entirely

TheEssem commented:
At the moment, when attempting to request statuses in bulk (e.g. for timelines), a separate database query is made for each status to get its reactions. This is in contrast to how favorites/reblogs/filters/etc. work, and is definitely not ideal for performance.

Seems like this is fixable by adding reactions to the status relationships (and adding a `reactions_map` function to the Status record class like the ones for favorites/reblogs/etc.), however I can't seem to figure out a good way to handle this right now...

TheEssem committed:
Disable reactions in detailed status view when visibleReactions is 0

TheEssem committed:
Disable reactions in detailed status view when visibleReactions is 0

kescherCode commented:
@EngineerDali This is a massive Pull Request, and changes that need to be made keep being found over time. The target branch also moves fast, as it has to keep up with vanilla Mastodon.

And yes, upstream changes break this Pull Request quite often. Right now, Essem does a great job at keeping up with that. Previously, three other maintainers of this patchset, including its original creator, have stopped managing to keep up with upstream changes, which is why this patchset is now in its **fourth** iteration.

All of these factors contribute to why this is not part of glitch-soc itself yet. I think we all hope it is fully reviewed and merged at some point, but as someone who maintained the prior iteration of this Pull Request, reviewed all of the code (I keep reviewing it everytime it changes as I maintain my fork of glitch-soc), I fully understand why it is not merged yet.

But things will eventually happen, I am sure in it.

ShadowJonathan reviewed (commented):
Putting some thoughts down while looking through this, I might merge it with tech.lgbt's fork if it doesnt merge with glitch-soc soon enough (maybe before/during summer vacation)

kescherCode commented on app/services/react_service.rb:
```diff
@@ -0,0 +1,46 @@
+    if status.account.local?
+      LocalNotificationWorker.perform_async(status.account_id, reaction.id, 'StatusReaction', 'reaction')
+    elsif status.account.activitypub?
+      ActivityPub::DeliveryWorker.perform_async(build_json(reaction), reaction.account_id, status.account.inbox_url)
```
Indeed, in Catstodon I modified this patch to allow incoming reacts to be saved even if they are for remote posts.

TheEssem committed:
Turn custom emoji regexps into class level constants

TheEssem committed:
Turn custom emoji regexps into class level constants

TheEssem commented on app/lib/activitypub/activity/emoji_react.rb:
```diff
@@ -0,0 +1,26 @@
+              !original_status.account.local? ||
+              delete_arrived_first?(@json['id'])
+
+    if /^:.*:$/.match?(name)
```
Fixed somewhat; there were a few other related occurrences of that regex that I ended up changing as well.

meghadeep-com commented:
Got a little impatient while waiting for this to be merged, and spun up a new personal instance (meghadeep.social) with this patchset. It's only been a day, and I have very little traffic (despite what the domain name suggests, I'm not very social), but I got no complaints so far. Huge shoutout to TheEssem  for taking this on! **Please let me know if I can help test anything.** 

If anybody else wants to have a go at it, I also pushed my podman container images on the docker hub (https://hub.docker.com/repository/docker/meghadeep/mastodon/tags), but please keep in mind that the images are manually generated and uploaded by me. so any subsequent updates will take time to get pushed out.

TheEssem commented:
This currently does not work with notification grouping. Currently attempting to add support, but it seems to be slightly more difficult than I expected.

TheEssem committed:
Add notification grouping for reactions

TheEssem committed:
Add notification grouping for reactions

TheEssem committed:
Fix reactions bar alignment in grouped notifications

TheEssem committed:
Fix reactions bar alignment in grouped notifications

gmemstr commented:
Looks like this needs another rebase :(

TheEssem committed:
Fix reblog reactions being hydrated improperly

TheEssem committed:
Fix reblog reactions being hydrated improperly

gmemstr commented:
Interestingly, applying this as a patch file to a fresh checkout of the main branch results in one conflict in `app/javascript/mastodon/locales/en.json`. After rebasing and fixing that conflict it applies cleanly.

I pasted my own rebased patch file here https://paste.gmem.ca/paste/0e0e6534-2bef-442f-8d73-5aff2a3514b4/raw - not sure if the PR itself needs a rebase+force push but I suspect that's the case.

TheEssem commented:
> Interestingly, applying this as a patch file to a fresh checkout of the main branch results in one conflict in `app/javascript/mastodon/locales/en.json`. After rebasing and fixing that conflict it applies cleanly.
> 
> I pasted my own rebased patch file here https://paste.gmem.ca/paste/0e0e6534-2bef-442f-8d73-5aff2a3514b4/raw - not sure if the PR itself needs a rebase+force push but I suspect that's the case.

I've gone ahead and done a manual rebase that should remove all references to files in `app/javascript/mastodon`, see if it works now.

TheEssem committed:
Fix grouped reaction notification text

TheEssem committed:
Fix grouped reaction notification text

TomCasavant commented:
@TheEssem Merged this into my server a few days ago, haven't had the chance to explore why this is happening so it might be something on my end, but the emoji picker in the web-ui seems to be offset from the rest of the buttons (glitch theme not vanilla)
![image](https://github.com/user-attachments/assets/31479b9c-3382-40eb-8d9d-cd62a6e25dac)


The emoji picker seems to show up in a <div> tag and if I move the button outside of that div it seems to fix it, but again this might be something I have misconfigured on my end. 
![image](https://github.com/user-attachments/assets/9ee63165-dc72-4bc3-ba1e-bde0bdefbc04)

TheEssem commented:
@TomCasavant Seems like your instance is using an old version of the Modern theme; try updating to the latest version, it should natively support this patch as long as you have the glitch fixes applied.

TomCasavant commented:
> @TomCasavant Seems like your instance is using an old version of the Modern theme; try updating to the latest version of the theme, it should natively support this patch as long as you have the glitch fixes applied.

Yup, sorry. Updating the theme and compiling the assets again fixed it

TheEssem committed:
Make addReaction and removeReaction optional props

This prevents things from breaking with embeds.

TheEssem committed:
Make addReaction and removeReaction optional props

This prevents things from breaking with embeds.

highlabs commented:
I'm running it on bantu.social and everything seems to be ok. :)

highlabs commented:
> I'm running it on bantu.social and everything seems to be ok. :)

Hey @TheEssem after testing your PR for a couple of days in my instances I collected some feedbacks from the users.
The reactions between Mastodon and other *keys instances (like sharkey) works like a charm. But on vanilla Mastodon instance (and probably others in the fediverse that don't support reactions) the users don't get any notifications. 
If I'm not wrong, on *keys instances, if you send a reaction to an instance that doesn't support it, the user on this instance will receive a "like".

kescherCode commented on app/mailers/notification_mailer.rb:
```diff
@@ -41,6 +41,14 @@ def favourite
   end
 
+  def reaction
+    return unless @user.functional? && @status.present?
```
```suggestion
    return if @status.blank?
```

TheEssem replied:
Fixed.

kescherCode reviewed (changes requested):
In all other notification types, but esp. in favorite, this has been changed to this.

TheEssem committed:
Align mail status check with upstream

TheEssem committed:
Align mail status check with upstream

TheEssem committed:
Fix invisible reactions on detailed statuses when logged out

TheEssem committed:
Fix invisible reactions on detailed statuses when logged out

TheEssem committed:
Fix status reactions not animating on hover when logged out

TheEssem committed:
Fix status reactions not animating on hover when logged out

TheEssem committed:
Update status reaction emails

TheEssem committed:
Update status reaction emails

TheEssem committed:
Fix old migration

TheEssem committed:
Fix old migration

kescherCode commented on app/serializers/initial_state_serializer.rb:
```diff
@@ -126,10 +131,15 @@ def default_meta_store
       trends_as_landing_page: Setting.trends_as_landing_page,
       trends_enabled: Setting.trends,
       version: instance_presenter.version,
+      visible_reactions: Setting.visible_reactions,
```
The most minor nitpick in the history of this pull request: This line should probably be moved to *after* `terms_of_service_enabled`, if alphabetical order is to be preserved.

TheEssem replied:
Perhaps, though in this case upstream didn't quite follow it either ("t" doesn't come after "v").

rigrig commented:
I tried merging this into [my fork](https://code.tubul.net/rigrig/glitch-soc) (glitch-soc with a few textual/cosmetic changes), but the migration failed:
```fish
SKIP_POST_DEPLOYMENT_MIGRATIONS=true bundle exec rails db:migrate --trace >>emoji_migrate.log 2>> emoji_migrate.log
```
I think the important part is 
```
PG::UndefinedColumn: ERROR:  column settings.thing_type does not exist
LINE 1: SELECT "settings".* FROM "settings" WHERE "settings"."thing_...
```
Full error log: [emoji_migrate.log](https://github.com/user-attachments/files/18304269/emoji_migrate.log)

kescherCode commented:
@rigrig You have be64f5f0e761fa0fb36c33d39dfc818a9558a854 applied too, right? If so, try reverting the changes made in that commit specifically (either by reverting the commit itself or making a reverse patch) and see if that makes a difference.

TheEssem commented:
That commit was specifically intended to work around that very issue, but it looks like it wasn't very effective...

rigrig commented:
I tried reverting be64f5f0e761fa0fb36c33d39dfc818a9558a854, but it still throws the same error.

kescherCode commented:
Hm... Does that migration even need to run for a fresh, initial merge of emoji reactions? @TheEssem

TheEssem commented:
> Hm... Does that migration even need to run for a fresh, initial merge of emoji reactions? @TheEssem

Not that I'm aware of, no; it could probably be removed in this case, it's mostly for compat with older iterations of the branch.

kescherCode commented:
In that case: @rigrigm, just remove the file for the migration for now

kescherCode commented:
In that case: @rigrig, just remove the file for the migration for now

TheEssem committed:
Remove emoji reaction settings move migration

It's now broken and pointless due to the complete removal of legacy user settings. Since this effectively predates the current reaction branch as well, there is little chance that someone running an older iteration will need to use this migration.

TheEssem committed:
Remove emoji reaction settings move migration

It's now broken and pointless due to the complete removal of legacy user settings. Since this effectively predates the current reaction branch as well, there is little chance that someone running an older iteration will need to use this migration.

rigrig commented:
Thanks! It seems to work now :tada:

gmemstr commented:
Looks like this patch needs another rebase:

```
Applying ./patches/0001-emoji-reactions.patch
Applying: Add support for emoji reactions Squashed, modified, and rebased from glitch-soc/mastodon#2221.
error: patch failed: app/serializers/rest/v1/instance_serializer.rb:89
error: app/serializers/rest/v1/instance_serializer.rb: patch does not apply
Patch failed at 0001 Add support for emoji reactions Squashed, modified, and rebased from glitch-soc/mastodon#2221.
```

TheEssem committed:
Group reaction notifications by default

TheEssem committed:
Group reaction notifications by default

gmemstr commented:
Beep boop, this needs another rebase :p 

Related, anyone have a good way of keeping a patch up to date without poking this PR to get rebased?

sinoru commented:
> Beep boop, this needs another rebase :p
> 
> Related, anyone have a good way of keeping a patch up to date without poking this PR to get rebased?

@gmemstr 
That could be challenging by the nature of Git. So… I hope this merged as soon as possible.
Mean while… How about using the [patch file](https://github.com/glitch-soc/mastodon/pull/2462.patch) to patch yourself on new base?

gmemstr commented:
> @gmemstr That could be challenging by the nature of Git. So… I hope this merged as soon as possible. Mean while… How about using the [patch file](https://github.com/glitch-soc/mastodon/pull/2462.patch) to patch yourself on new base?

I am using a patch file :) I have this repo containing the patches and build scripts https://github.com/floofy-mastodon/floofy.tech

minacle commented:
[mastodon-bd963c7-Receive reaction from other instance for another instances.patch](https://github.com/user-attachments/files/19072190/mastodon-bd963c7-Receive.reaction.from.other.instance.for.another.instances.patch)

This patch I created allows users from one instance to see reactions from users of another instance.

sinoru commented:
> [mastodon-bd963c7-Receive reaction from other instance for another instances.patch](https://github.com/user-attachments/files/19072190/mastodon-bd963c7-Receive.reaction.from.other.instance.for.another.instances.patch)
> 
> This patch I created allows users from one instance to see reactions from users of another instance.

For the context explanation:
Originally, Mastodon only interprets `like` activities for own instance's posts. But this patch will loosen that limitation because, Emoji reactions are more complex than `like` which should be accept as possible for other instances' posts.
(Emoji reactions is attached to `like` activity)

kescherCode commented:
@minacle This is only a partial patch: you might want to apply similar changes to `emoji_react.rb` from this PR.

minacle commented:
@kescherCode

> @minacle This is only a partial patch: you might want to apply similar changes to `emoji_react.rb` from this PR. Also, `undo.rb` needs further changes for `undo_emoji_reaction`.

Thank you for pointing it out. As you mentioned, I have made changes to `emoji_react.rb` and also modified `undo.rb` for `undo_emoji_reaction`. I am now posting a new patch with these updates.

[mastodon-e9d7565-Receive reaction from other instance for another instances.patch](https://github.com/user-attachments/files/19141627/mastodon-e9d7565-Receive.reaction.from.other.instance.for.another.instances.patch)

This patch allows users from one instance to see reactions from users of another instance.

sinoru commented:
> [mastodon-e9d7565-Receive reaction from other instance for another instances.patch](https://github.com/user-attachments/files/19141627/mastodon-e9d7565-Receive.reaction.from.other.instance.for.another.instances.patch)
> 
> This patch allows users from one instance to see reactions from users of another instance.

@TheEssem How do you think about cherry pick [this patch](https://github.com/user-attachments/files/19141627/mastodon-e9d7565-Receive.reaction.from.other.instance.for.another.instances.patch) to this PR?

TheEssem commented:
The reason why I haven't merged the patch into this PR is because I aim to follow upstream's behavior when it comes to likes, which are probably the closest existing type of interaction to reactions.

That being said, I might consider merging it into my own fork along with a similar patch for likes, since it's been a common suggestion and I don't have to care as much about upstream behavior there.

agaricus-campestris commented:
Looking for emoji reaction on glitch, looks amazing ! When / in what version this would be merged ?

o (◡‿◡✿)

Just in case, [I asked in moshidon issues if this would be possible to get this feature](https://github.com/LucasGGamerM/moshidon/issues/614). Thanks a bunch for all the work !

kescherCode commented:
@agaricus-campestris Moshidon already supports emoji reactions, you just need to enable the feature yourself in the app settings per profile (because currently, this PR exposes no way to show emoji reactions exist)

TheEssem committed:
Migrate status reactions to react-spring

TheEssem committed:
Migrate status reactions to react-spring

TheEssem committed:
Do not fade in emoji reactions on initial load

TheEssem committed:
Do not fade in emoji reactions on initial load

agaricus-campestris commented:
Hi, I didn't understand how to use this emoji reaction feature onto a glitch-soc last update server. Sorry for being very noobish here, using tangerine UI too here, it seems I couldn't make this work from Moshidon too (got a 404). 

Thanks a bunch for all the work.

TheEssem committed:
Add toot:Emoji and litepub:EmojiReact context extensions to reaction objects

TheEssem committed:
Add toot:Emoji and litepub:EmojiReact context extensions to reaction objects

TheEssem committed:
Fix reaction notifications in web UI

TheEssem committed:
Fix reaction notifications in web UI

TheEssem committed:
Add reaction to push subscriptions spec

TheEssem committed:
Add reaction to push subscriptions spec

TheEssem commented on app/javascript/flavours/glitch/components/status_action_bar.jsx:
```diff
@@ -225,6 +233,8 @@ class StatusActionBar extends ImmutablePureComponent {
     this.props.onFilter();
   };
 
+  handleNoOp = () => {}; // hack for reaction add button
```
Indeed it is, removing.

TheEssem committed:
Remove unused no-op handler

TheEssem committed:
Remove unused no-op handler

TheEssem committed:
Do not use fixed precision in reaction schema

TheEssem committed:
Do not use fixed precision in reaction schema

DarkCat09 commented:
@derutaramuda 

https://github.com/glitch-soc/mastodon/pull/2462#issuecomment-2174429521

TheEssem committed:
Fix color of reaction notif icon

TheEssem committed:
Fix color of reaction notif icon

TheEssem committed:
Fix visual bug when number of visible reactions is 0

TheEssem committed:
Fix visual bug when number of visible reactions is 0

TheEssem committed:
Consolidate title/icon props into single boolean

TheEssem committed:
Consolidate title/icon props into single boolean

TheEssem committed:
Switch from Oj to JSON

TheEssem committed:
Switch from Oj to JSON

TheEssem committed:
Use to_json in unreact service

TheEssem committed:
Use to_json in unreact service

TheEssem committed:
Fix potential leak of whether a private post exists in reactions controller

GHSA-gwhw-gcjx-72v8 seems to affect this as well. Let's patch that up.

TheEssem committed:
Fix potential leak of whether a private post exists in reactions controller

GHSA-gwhw-gcjx-72v8 seems to affect this as well. Let's patch that up.

TheEssem committed:
Add tests for status reaction endpoint

TheEssem committed:
Add tests for status reaction endpoint

TheEssem committed:
Fix removing reactions from blocked posts

TheEssem committed:
Fix removing reactions from blocked posts

TheEssem committed:
Fix notification mail subject

TheEssem committed:
Fix notification mail subject

TheEssem committed:
Add reaction notification mail test

TheEssem committed:
Add reaction notification mail test

TheEssem committed:
Add tests for emoji react federation

TheEssem committed:
Add tests for emoji react federation

TheEssem committed:
Add tests for emoji react undos

TheEssem committed:
Add tests for emoji react undos

quagsirus commented on app/javascript/flavours/glitch/components/status.jsx:
```diff
@@ -794,4 +810,4 @@ class Status extends ImmutablePureComponent {
 }
 
-export default withOptionalRouter(injectIntl(Status));
+export default withOptionalRouter(injectIntl((withIdentity(Status))));
```
the addition of `withIdentity` here is causing the typecheck error, that function is stripping props

i wrote a fix in https://github.com/TheEssem/mastodon/pull/53

quagsirus replied:
copied errors from [the ci job](https://github.com/glitch-soc/mastodon/actions/runs/24966634516/job/73102472304?pr=2462) in case it expires or something
```
Error: app/javascript/flavours/glitch/components/status_quoted.tsx(339,8): error TS2322: Type '{ children: false | Element | undefined; isQuotedPost: true; id: string | null | undefined; contextType: string | undefined; avatarSize: number; headerRenderFn: StatusHeaderRenderFn; }' is not assignable to type 'IntrinsicAttributes & { ref: unknown; wrappedComponentRef: unknown; contextType: any; context?: Context<ReactReduxContextValue<any, UnknownAction> | null> | undefined; store?: Store<...> | undefined; }'.
  Property 'children' does not exist on type 'IntrinsicAttributes & { ref: unknown; wrappedComponentRef: unknown; contextType: any; context?: Context<ReactReduxContextValue<any, UnknownAction> | null> | undefined; store?: Store<...> | undefined; }'.
Error: Process completed with exit code 2.
```

TheEssem committed:
Adjust for new emoji system

TheEssem committed:
Adjust for new emoji system

TheEssem committed:
Custom emoji/animations fixes

TheEssem committed:
Custom emoji/animations fixes

TheEssem committed:
Fix reaction reply styling

TheEssem committed:
Fix reaction custom emoji rendering

TheEssem committed:
Normalize unicode emoji variation selectors

999eagle commented on db/post_migrate/20260520184138_normalize_status_reaction_variation_selectors.rb:
```diff
@@ -0,0 +1,13 @@
+  class StatusReaction < ApplicationRecord; end
+
+  def up
+    StatusReaction.where(custom_emoji_id: nil).find_each do |react|
```
this migration fails executing on here with this log:

```
I, [2026-05-20T23:03:05.103059 #2242526]  INFO -- : Migrating to NormalizeStatusReactionVariationSelectors (20260520184138)                               
== 20260520184138 NormalizeStatusReactionVariationSelectors: migrating ========                                                                           
bin/rails aborted!                                                                                                                                        
StandardError: An error has occurred, this and all later migrations canceled: (StandardError)                                                             
PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_status_reactions_on_account_id_and_status_id"                          
DETAIL:  Key (account_id, status_id, name)=([redacted], [redacted], ❤️) already exists.          
ActiveRecord::RecordNotUnique: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_status_reactions_on_account_id_and_status_id" (ActiveRecord::RecordNotUnique)
```

looking in the database, apparently there are 25 combinations of account_id/status_id that have both a '❤️' and a '❤' reaction (i.e. emoji representation and text representation), which should probably be reduced to only the emoji representation of the reaction, either manually or through a fixed version of this migration

kescherCode replied:
Having also encountered this, I have written a migration that seems to finally be working.
It is also painfully slow, but given this is a post-deployment migration, that shouldn't be too bad either.
I will be making a PR on the source branch of this PR soon

kescherCode replied:
Running post-deployment migrations...
== 20260520184138 NormalizeStatusReactionVariationSelectors: migrating ========
== 20260520184138 NormalizeStatusReactionVariationSelectors: migrated (735.1358s)

that sure took a while lmao. anyway

kescherCode replied:
anyway https://github.com/TheEssem/mastodon/pull/54

TheEssem committed:
Fix custom emoji regex again

TheEssem committed:
Remove identity context from Status

It's unused now and causing issues. Get rid of it.
